nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-distributedmapcache-service/src/main/java/org/apache/nifi/controller/cassandra/CassandraDistributedMapCache.java

```diff
@@ -0,0 +1,241 @@
+                    "old cache items expire after a certain period of time.")
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .required(false)
+            .build();
+
+    public static final List<PropertyDescriptor> DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(
+        SESSION_PROVIDER, TABLE_NAME, KEY_FIELD_NAME, VALUE_FIELD_NAME, TTL
+    ));
+
+    private CassandraSessionProviderService sessionProviderService;
+    private String tableName;
+    private String keyField;
+    private String valueField;
+    private Long ttl;
+
+    private Session session;
+    private PreparedStatement deleteStatement;
+    private PreparedStatement existsStatement;
+    private PreparedStatement fetchStatement;
+    private PreparedStatement insertStatement;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return DESCRIPTORS;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        sessionProviderService = context.getProperty(SESSION_PROVIDER).asControllerService(CassandraSessionProviderService.class);
+        tableName = context.getProperty(TABLE_NAME).evaluateAttributeExpressions().getValue();
+        keyField = context.getProperty(KEY_FIELD_NAME).evaluateAttributeExpressions().getValue();
+        valueField = context.getProperty(VALUE_FIELD_NAME).evaluateAttributeExpressions().getValue();
+        if (context.getProperty(TTL).isSet()) {
+            ttl = context.getProperty(TTL).evaluateAttributeExpressions().asTimePeriod(TimeUnit.SECONDS);
+        }
+
+        session = sessionProviderService.getCassandraSession();
+
+        deleteStatement = session.prepare(createDeleteStatement(keyField, tableName));
+        existsStatement = session.prepare(createExistsQuery(keyField, tableName));
+        fetchStatement = session.prepare(createFetchQuery(keyField, valueField, tableName));
+        insertStatement = session.prepare(createInsertStatement(keyField, valueField, tableName, ttl));
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        session = null;
+        deleteStatement = null;
+        existsStatement = null;
+        fetchStatement = null;
+        insertStatement = null;
+    }
+
+    @Override
+    public <K, V> boolean putIfAbsent(K k, V v, Serializer<K> keySerializer, Serializer<V> valueSerializer) throws IOException {
+        if (containsKey(k, keySerializer)) {
+            return false;
+        } else {
+            put(k, v, keySerializer, valueSerializer);
+            return true;
+        }
+    }
+
+    @Override
+    public <K, V> V getAndPutIfAbsent(K k, V v, Serializer<K> keySerializer, Serializer<V> valueSerializer, Deserializer<V> deserializer) throws IOException {
+        V got = get(k, keySerializer, deserializer);
+        boolean wasPresent = putIfAbsent(k, v, keySerializer, valueSerializer);
```

Comment:
Shouldn't this be the inverse? `putIfAbsent` will return false if the value was present in the cache, so `wasPresent` should be true. The HBase versions do:

```
final boolean wasAbsent = putIfAbsent(key, value, keySerializer, valueSerializer);

if (! wasAbsent) return got;
else return null;
```
Would be good to add a unit test around this as well.

Reply:
Good catch.

Reply:
Should be fixed now.